CONTRIBUTING: Add note about how to commit new maintainers-list.nix entry

matthiasbeyer opened the pull request:
In #247972 I found that the contributing file does not say how to add a maintainer, only the manual does (as @ius said in above PR).

Maybe we can add a note here.

matthiasbeyer committed:
CONTRIBUTING: Add note about how to commit new maintainers-list.nix entry

Signed-off-by: Matthias Beyer <mail@beyermatthias.de>

Janik-Haag commented:
this probably relates to #245243

ius reviewed (dismissed):
Thanks, beat me to it.

Two remarks:

1. `<your github handle>`: just `handle` I think, see [this remark](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L21) - it doesn't have to be a GitHub username (not all GH names are valid attribute anyway, I think?)

2. Related: I wonder whether it should be explicit about submitting it as a separate commit. I think it makes sense and I see many reviewers request this, but on the other hand most committers/the GitHub default end up creating a merge commit, thus hiding said commit?

matthiasbeyer commented:
Fixed what you suggested!

I will squash this if someone requests it.

infinisil commented:
> this probably relates to #245243

Thanks for spreading awareness, I'd love some reviews for that once it's not a draft anymore btw, feel free to subscribe to it!

That PR is still a bit too much in flux though, I don't think it should be blocking this PR here. I can rebase my PR once this one is merged (yeah it's a bit annoying to rebase when things have been moved, but I can manage :wink:)

Lord-Valen commented:
> 2. Related: I wonder whether it should be explicit about submitting it as a separate commit. I think it makes sense and I see many reviewers request this, but on the other hand most committers/the GitHub default end up creating a merge commit, thus hiding said commit?

I think it should be. Also, merge commits don't hide other commits if what you mean by "hide" is something that implies that they somehow aren't in the branch. A merge commit adds a new commit with two parent commits (the latest commit of each branch) that resolves any conflicts. Both parent commits are included in the receiving branch. You might think that merge commits hide other commits since the dates of those commits don't change and the merge commit diff shows the same changes. A merge commit can occur any amount of time after its parents, which can result in a large distance between them. With the velocity of Nixpkgs, that usually means the merge commit and one of the parents have quite a few commits between them. Here's an example: https://github.com/NixOS/nixpkgs/commit/195975b1387b69ac5566338819298d7abc595ac3 merges  https://github.com/NixOS/nixpkgs/commit/abc3cc0ac6f44ebbd997464c2e1460c9001d7743 into master on https://github.com/NixOS/nixpkgs/commit/e04428a3a8e4e672314011c07d56c18b62e34e89.
Right now there are 22 commits on master in the time between https://github.com/NixOS/nixpkgs/commit/195975b1387b69ac5566338819298d7abc595ac3 and https://github.com/NixOS/nixpkgs/commit/abc3cc0ac6f44ebbd997464c2e1460c9001d7743.

If you want to see an example of a merge commit for a PR with an additional `maintainers:` commit, you could explore the parents of https://github.com/NixOS/nixpkgs/commit/36ea4a4b6ab950c3ea1ea9d7fe036f35bd85d2cb.

Lord-Valen commented on CONTRIBUTING.md:
```diff
@@ -61,6 +61,8 @@ Pull requests should not be squash merged in order to keep complete commit messa
 This means that, when addressing review comments in order to keep the pull request in an always mergeable status, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
 Useful git commands that can help a lot with this are `git commit --patch --amend` and `git rebase --interactive`. For more details consult the git man pages or online resources like [git-rebase.io](https://git-rebase.io/) or [The Pro Git Book](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).
 
+If you add yourself as maintainer, use `maintainers: add <handle>` as commit message ([see also](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers).
```
Yes, I would put the commit message format in as an example and change the template accordingly under "Submitting Changes" and also try to make it more obvious that changes to `maintainer-list.nix` should occur in a separate commit.

I think it might be reasonable to also normalize `doc/<page>:` and `lib/<module>:` commit messages or simply `doc:` and `lib:` for more general changes, but that isn't necessary here.

infinisil replied:
Ah yeah then I guess this fits very well into the scope of the contributing documentation cleanup in #245243, because I create separate contributing documents for each part of Nixpkgs, such as `pkgs/README.md`, `doc/README.md`, `maintainers/README.md`, etc. And this would fit very well into `maintainers/README.md`.

matthiasbeyer replied:
I added the change requested by @infinisil as a commit to this PR, which I can squash if requested of course.

AndersonTorres commented on CONTRIBUTING.md:
```diff
@@ -37,6 +37,8 @@ Below is a short excerpt of some points in there:
 
     The old config generation system used impure shell scripts and could break in specific circumstances (see #1234).
 
+  If you add yourself as maintainer, use `maintainers: add <handle>` as commit message ([see also](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers) and as a seperate commit.
```
```suggestion
  In case of adding yourself as maintainer, do it in a separated commit (no need for a Pull Request), with message `maintainers: add <handle>`. See also ([the section of manual about meta.maintainers](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers).
```

Lord-Valen replied:
It isn't even necessarily adding yourself. It's adding anyone to the maintainer list.

AndersonTorres replied:
This is the most common scenario: a newcomer adding a package.
But point taken, I will update this.

fricklerhandwerk commented on CONTRIBUTING.md:
```diff
@@ -37,6 +37,8 @@ Below is a short excerpt of some points in there:
 
     The old config generation system used impure shell scripts and could break in specific circumstances (see #1234).
 
+  In case of adding yourself as maintainer, do it in a separated commit (no need for a Pull Request), with message `maintainers: add <handle>`. See also ([the section of manual about meta.maintainers](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers).
```
```suggestion
  When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
  See [Nixpkgs Maintainers](../maintainers/README.md) for details.
```

Given the information will be expanded in https://github.com/NixOS/nixpkgs/pull/250344, we should rather link to that file. We can still add information on NixOS contributions later.

VuiMuich commented:
On my PR adding myself to `maintainers` I got advised to reorder the commits, so it is before the package commit. Might be worth adding a side note for this being preferred. 
Idk if this technically makes any CI checks easier (fail earlier in eval maybe?), but logically it makes sense at least.

fricklerhandwerk commented on CONTRIBUTING.md:
```diff
@@ -37,6 +37,9 @@ Below is a short excerpt of some points in there:
 
     The old config generation system used impure shell scripts and could break in specific circumstances (see #1234).
 
+  When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
```
```suggestion
  When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
Add the commit before those making changes to the package or module.
```

VuiMuich replied:
Is it intentional, or is the indentation off?

fricklerhandwerk replied:
Ech, no, that was a typo.

fricklerhandwerk reviewed (approved):
@matthiasbeyer please rebase.

infinisil reviewed (changes requested):
This is in the wrong place now, should be in `maintainers/README.md`

fricklerhandwerk commented:
@Infinisil: this section is about commit formatting. The organizational details are in the other file. I think it makes sense to have it here, and also use the opportunity to direct people to the new extended page.

infinisil commented:
True there are commit conventions in `CONTRIBUTING.md`, but they're [further down](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions). I'd be fine with it there, but I do think it should be in `maintainers/README.md` in the future (and same for package commit conventions, which should be in `pkgs/README.md`). But the PR here currently adds it to the PR template section, which is definitely wrong.

VuiMuich commented:
I agree, that it would be a better fit in the "How to create pull requests" section, probably as part of bullte 4.

Imo it would not hurt to double this small information in both the `CONTRIBUTING.md` and `maintainers/README.md` as it is easy to be overlooked.

infinisil commented:
@VuiMuich I'd rather not have it be duplicated. It makes it harder to know which part should be updated, or if you don't know about both parts they could get outdated. I think it's better to improve the linking between the sections.

> I agree, that it would be a better fit in the "How to create pull requests" section, probably as part of bullet 4.

Step 4 links to the commit conventions [here](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions), where there's already information about the commit message formatting.

cafkafk commented on CONTRIBUTING.md:
```diff
@@ -155,6 +155,10 @@ Many Nix packages are designed to run on multiple platforms. As such, it’s imp
 
 Packages with automated tests are much more likely to be merged in a timely fashion because it doesn’t require as much manual testing by the maintainer to verify the functionality of the package. If there are existing tests for the package, they should be run to verify your changes do not break the tests. Tests can only be run on Linux. For more details on writing and running tests, see the [sect ...
 
+When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
```
```suggestion
When adding yourself as maintainer, in the same pull request, make a separate commit with the message `maintainers: add <handle>`.
```

cafkafk reviewed (approved):
Comma nitpick but lgtm

VuiMuich commented:
I can follow that "harder to update" argument very well and as I have stumbled on a bunch of such discrepancies between different docs in the bigger nix eco system. With this in mind, I agree now, to better avoid this duplication.

Also putting it as a bullet in the "commit conventions" totally makes sense.

fricklerhandwerk commented:
@matthiasbeyer can you follow up with the changes? Since there seems to be a lot of energy behind this right now, if you lack time the docs team can take this over.

matthiasbeyer commented:
I was traveling today, I can have a look tomorrow!

matthiasbeyer commented:
Squashed the comma commit from @cafkafk and added a new one moving the text block to "commit conventions" as per request.

I can squash these commits as well, if you like!

cafkafk commented:
It probably makes most sense to have just a single commit for this, instead of one adding it and another moving it.

asymmetric commented:
Not sure if here or in another PR, but it would be good to also document how to add oneself to `meta.maintainers`, meaning just what you did in this PR:

- how the commit should be formatted
- the placement (together with a commit making a change to a package, or separately?)

Same for removal probably.

infinisil commented:
@asymmetric It's mostly there in `maintainers/README.md`, though it could use a bit of work

fricklerhandwerk reviewed (approved):
Let's merge this first and follow up afterwards.
